python/mooseutils/ImageDiffer.py

```diff
@@ -0,0 +1,173 @@
+        for e in self.__errors:
+            cnt += 1
+            output.append('')
+            output.append('ERROR ' + str(cnt) + ':')
+            output.append('  ' + e[0])
+            if e[1]:
+                output.append('    ' + e[1])
+
+        # Print the output
+        if kwargs.pop('output', False):
+            print '\n'.join(output)
+
+        # Return the text, as a single string
+        return '\n'.join(output)
+        # Errors
+        cnt = 0
+        for e in self.__errors:
+            cnt += 1
+            output.append('')
+            output.append('ERROR ' + str(cnt) + ':')
+            output.append('  ' + e[0])
+            if e[1]:
+                output.append('    ' + e[1])
+
+        # Print the output
+        if kwargs.pop('output', False):
+            print '\n'.join(output)
+
+        # Return the text, as a single string
+        return '\n'.join(output)
+
+    def __compare(self):
+        """
+        Perform image comparison. (private)
+        """
+
+        # Do nothing if something failed to open
+        if len(self.__errors) > 0:
+            return
+
+        # Check sizes
+        if (self.__data[0].size != self.__data[1].size):
+            err = 'The two images are different sizes'
+            msg  =  ['  File 1: ' + self.__files[0]]
+            msg +=  ['    size: ' + str(self.__data[0].size)]
+            msg +=  ['  File 2: ' + self.__files[1]]
+            msg +=  ['    size: ' + str(self.__data[1].size)]
+            self.__addError(err, msg)
+            return
+
+        # Compute the error
+        self.__error = skimage.measure.compare_ssim(self.__data[0], self.__data[1], multichannel=True)
+
+        # Report the error
+        if self.__error < self.__allowed:
+            err = 'The files are different.'
+            msg = ['The difference of the images exceeds the "allowed" SSIM.']
+            msg += ['                 Allowed (SSIM): ' + str(self.__allowed)]
+            msg += ['                Computed (SSIM): ' + str(self.__error)]
+            msg += ['                Rel. difference: ' + str( abs(self.__allowed - self.__error) / self.__error)]
+
+            self.__addError(err, msg)
+            return
+
+    def __readImage(self, filename):
+        """
+        A read function that appends an error message if the read fails. (private)
+        """
+
+        if not os.path.exists(filename):
+            self.__addError('Failed to open ' + filename + ', the file does not exist.')
+            return None
+
+        return plt.imread(filename)
+
+    def __addError(self, err, msg=[]):
+        """
+        Add an ImageError object to the storage vector (private)
+
+        Args:
+          err[str]: A string containing the error message.
+          msg[list]: A detailed message for the error.
+        """
+        self.__errors.append((err, '\n'.join(msg)))
+
+
+# This file is executable and performs tests of the functionality
+if __name__ == '__main__':
```

Comment:
This isn't really a test, despite what the comment says.
It just allows for doing `ImageDiffer.py image1.png image2.png`
We currently don't have any unittests for ImageDiffer, although obviously we have a ton of tests that use it. 
Do you want to add tests in this PR (along with the gold images that this would probably require)?